app/controllers/application_controller.rb

```diff
@@ -1,10 +1,133 @@
 require './config/environment'
 
+# ApplicationController handles requests for homepage, signup and logins
 class ApplicationController < Sinatra::Base
 
   configure do
     set :public_folder, 'public'
     set :views, 'app/views'
+
+    enable :sessions
+    set :session_secret, "secret"
+  end
+
+  get '/' do
+    redirect "/tweets" if Helpers.is_logged_in?(session)
+    erb :index
+  end
+
+  get '/signup' do
+    redirect "tweets" if Helpers.is_logged_in?(session)
+    erb :"users/signup"
+  end
+
+  post '/signup' do
+    if params[:username] == "" || params[:email] == "" || params[:password] == ""
+      redirect "/signup", locals: {message: "Please submit enter all the information to complete signup."}
+    else
+      @user = User.create(
+        username: params[:username],
+        email: params[:email],
+        password: params[:password]
+      )
+      session[:user_id] = @user.id
+      redirect "/tweets", locals: {message: "Welcome to Fwitter. Thank you for signing up."}
+    end
+  end
+
+  get '/login' do
+    redirect '/tweets'if Helpers.is_logged_in?(session)
+    erb :"users/login"
+  end
+
+  post '/login' do
+    if params[:username] == "" || params[:password] == ""
+      redirect "/login"
+    end
+
+    @user = User.find_by(username: params[:username])
+    if  @user && @user.authenticate(params[:password])
+      session[:user_id] = @user.id
+      redirect "/tweets"
+    else
+      redirect "/login"
+    end
+  end
+
+  get '/logout' do
+    session.clear if Helpers.is_logged_in?(session)
+    redirect "/login" if !Helpers.is_logged_in?(session)
+  end
+
+  get '/users' do
```

Comment:
This controller is overloaded. Consider separating your concerns by splitting some of these actions into separate controllers (i.e. a users controller and a tweets controller).